File: dsa-python/solutions/linked_list/medium/lru_cache.py
# 146. LRU Cache
# https://leetcode.com/problems/lru-cache/
#
# Design a data structure that follows the constraints of a Least Recently Used (LRU) cache.
#
# APPROACH: HashMap + Doubly Linked List.
# HashMap: key → Node for O(1) lookup.
# Doubly Linked List: maintains recency order (head = most recent, tail = least recent).
# Dummy head/tail avoid null checks when adding/removing.
#
# get: lookup in map, move to front, return value.
# put: if exists remove old, create new node, add to front, evict tail if over capacity.
#
# TIME: O(1) for both get and put
# SPACE: O(capacity) — for storing cache entries


class Node:
    def __init__(self, key: int, val: int):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None


class LRUCache:
    def __init__(self, capacity: int):
        self.map = {}
        self.capacity = capacity
        self.head = Node(0, 0)  # dummy head (most recent side)
        self.tail = Node(0, 0)  # dummy tail (least recent side)
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        if key not in self.map:
            return -1
        node = self.map[key]
        self._remove(node)
        self._add_to_front(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if key in self.map:
            self._remove(self.map[key])

        node = Node(key, value)
        self._add_to_front(node)
        self.map[key] = node

        if len(self.map) > self.capacity:
            lru = self.tail.prev  # least recently used
            self._remove(lru)
            del self.map[lru.key]  # need key to remove from map!

    def _remove(self, node):
        # Unlink node from its current position
        node.prev.next = node.next
        node.next.prev = node.prev

    def _add_to_front(self, node):
        # Insert right after dummy head
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node

File: dsa-csharp/solutions/LinkedList/Medium/LRUCache.cs
// 146. LRU Cache
// https://leetcode.com/problems/lru-cache/
//
// Design a data structure that follows the constraints of a Least Recently Used (LRU) cache.
//
// Implement the LRUCache class:
// - LRUCache(capacity) Initialize the LRU cache with positive size capacity.
// - Get(key) Return the value of the key if the key exists, otherwise return -1.
// - Put(key, value) Update the value of the key if the key exists. Otherwise, add the
//   key-value pair to the cache. If the number of keys exceeds the capacity, evict the
//   least recently used key.
//
// The functions Get and Put must each run in O(1) average time complexity.
//
// CATEGORY: Linked List (Medium)
//
// HINTS:
// - Use a Dictionary + LinkedList (or custom doubly linked list).
// - Dictionary maps key → node for O(1) lookup.
// - Doubly linked list maintains order (most recent at head, least recent at tail).
// - On Get/Put, move the node to the head.
// - On Put (over capacity), remove the tail node.
//
// TIME: O(1) for both Get and Put
// SPACE: O(capacity) — for storing cache entries

using System.Data;

public class LRUCache
{
    private class DoubleNode
    {
            public int Key;
            public int Val;
            public DoubleNode? Prev;
            public DoubleNode? Next;
            public DoubleNode(int key, int val)
        {
            Key = key;
            Val = val;
        }
    }

    Dictionary<int, DoubleNode> map;
    private DoubleNode head;
    private DoubleNode tail;
    private int capacity;
    public LRUCache(int capacity) //Head ->DoublyLinkedList <- Tail
    {
        this.capacity = capacity;
        map = new Dictionary<int, DoubleNode>();
        head = new DoubleNode(0,0);
        tail = new DoubleNode(0,0);
        head.Next = tail;
        tail.Prev = head;
    }

    public int Get(int key)
    {
        if(!map.ContainsKey(key)) return -1;
        DoubleNode node = map[key];
        Remove(node);
        AddToFront(node);
        return node.Val;
        
    }

    public void Put(int key, int value)
    {
        if(map.ContainsKey(key))
        {
            Remove(map[key]);
        }

        DoubleNode node = new DoubleNode(key, value);
        AddToFront(node);
        map[key] = node;

        if(map.Count > capacity)
        {
            DoubleNode lru = tail.Prev;
            Remove(lru);
            map.Remove(lru.Key); //need key to be saved in node to be removed from map
        }
    }

    private void Remove(DoubleNode node)
    {
        //Unlink node from it's current position
        node.Prev.Next = node.Next;
        node.Next.Prev = node.Prev;
        
    }
    private void AddToFront(DoubleNode node)
    {
        //Insert right after dummy head
        node.Next = head.Next;
        node.Prev = head;
        head.Next.Prev = node;
        head.Next = node;
    }
}


